runtime/near-vm-runner/src/tests/runtime_errors.rs

```diff
@@ -873,3 +883,295 @@ fn test_nan_sign() {
+                loading_gas,
+            );
+            gas_and_error_match(
+                res,
+                loading_gas,
+                Some(VMError::FunctionCallError(FunctionCallError::HostError(
+                    HostError::GasExceeded,
+                ))),
+            );
+        });
+    }
+
+    /// Executing with enough gas to finish loading but not to execute the full
+    /// contract should have the same outcome before and after.
+    #[test]
+    fn test_fn_loading_gas_protocol_upgrade_exceed_executing() {
+        with_vm_variants(|vm_kind: VMKind| {
+            let (code, exec_gas) = almost_trivial_contract_and_exec_gas();
+            let loading_gas = prepaid_loading_gas(code.len());
+            let total_gas = exec_gas + loading_gas;
+
+            let res = make_simple_contract_call_ex(
+                &code,
+                "main",
+                vm_kind,
+                Some(ProtocolFeature::FixContractLoadingCost.protocol_version() - 1),
+                total_gas - 1,
+            );
+            gas_and_error_match(
+                res,
+                total_gas - 1,
+                Some(VMError::FunctionCallError(FunctionCallError::HostError(
+                    HostError::GasExceeded,
+                ))),
+            );
+            let res = make_simple_contract_call_ex(
+                &code,
+                "main",
+                vm_kind,
+                Some(ProtocolFeature::FixContractLoadingCost.protocol_version()),
+                total_gas - 1,
+            );
+            gas_and_error_match(
+                res,
+                total_gas - 1,
+                Some(VMError::FunctionCallError(FunctionCallError::HostError(
+                    HostError::GasExceeded,
+                ))),
+            );
+        });
+    }
+
+    fn function_not_defined_contract() -> Vec<u8> {
+        wat::parse_str(
+            r#"
+            (module
+              (export "hello" (func 0))
+            )"#,
+        )
+        .unwrap()
+    }
+
+    #[track_caller]
+    fn check_failed_compilation(vm_kind: VMKind, code: &[u8], expected_error: CompilationError) {
+        let new_version = ProtocolFeature::FixContractLoadingCost.protocol_version();
+        let old_version = new_version - 1;
+        {
+            let res = make_simple_contract_call_with_protocol_version_vm(
+                &code,
+                "main",
+                old_version,
+                vm_kind,
+            );
+            // This is the first property the tests check: Gas cost before
+            // compilation errors must remain zero for old versions.
+            let expected_gas = 0;
+            gas_and_error_match(
+                res,
+                expected_gas,
+                Some(VMError::FunctionCallError(FunctionCallError::CompilationError(
+                    expected_error.clone(),
+                ))),
+            );
+        }
+
+        {
```

Comment:
I'd just split this into two `#[test]`s. There's little shared setup here and having more granular tests makes it easier to see which one fails when they do fail.

Reply:
I am not quite understanding the suggestion.

I mean, yes, I can easily split it into two checker functions. But how to split the actual test `test_fn_loading_gas_protocol_upgrade_fail_preparing`? In one way or another, I will have to duplicate the list of (contract,expected_error) pairs. I'm unable to make this work without having horrible duplication at one place or another... But maybe I just misunderstood the idea.

Reply:
Aha, never mind, I missed that this is a helper function not an actual test.